### PR TITLE
[cxx-interop] Add `CxxRandomAccessCollection` protocol

### PR DESCRIPTION
**Pull request by egorzhdan**
This helps to bridge C++ random access collections, such as `std::vector` and `std::string`, to Swift by conforming them to `Swift.RandomAccessCollection`.

### CODE REVIEW

**Comment by egorzhdan** [comment]
@swift-ci please smoke test

**Comment by egorzhdan** [comment]
@swift-ci please smoke test

**Comment by egorzhdan** [comment]
@swift-ci please smoke test macOS

**Review by zoecarver — APPROVED** [review]
This is really great. As always, a very high quality patch. Thanks Egor!

**Comment by zoecarver** [comment]
It will be interesting to see how this changes any std::vector benchmarks (maybe we should add some more).

**Comment by egorzhdan** [comment]
@swift-ci please benchmark

**Comment by hyp** [comment]
benchmark is disabled for now so you won't get results

File: stdlib/public/Cxx/CxxRandomAccessCollection.swift [hunk 1]
```diff
@@ -0,0 +1,65 @@
...
+  }
+
+  public var count: Int {
+    return Int(__endUnsafe() - __beginUnsafe())
+  }
+
+  /// A C++ implementation of the subscript might be more performant. This 
+  /// overload should only be used if the C++ type does not define `operator[]`.
+  public subscript(_ index: Int) -> Element {
+    get {
```
Comment (hyp) [inline stdlib/public/Cxx/CxxRandomAccessCollection.swift:58]: would it be possible to provide a `_read` accessor that can actually yield the reference?

File: stdlib/public/Cxx/CxxRandomAccessCollection.swift [hunk 2]
```diff
@@ -0,0 +1,65 @@
...
+    return count
+  }
+
+  public var count: Int {
+    return Int(__endUnsafe() - __beginUnsafe())
+  }
+
+  /// A C++ implementation of the subscript might be more performant. This 
+  /// overload should only be used if the C++ type does not define `operator[]`.
+  public subscript(_ index: Int) -> Element {
```
Comment (hyp) [inline stdlib/public/Cxx/CxxRandomAccessCollection.swift:57]: make it inlinable?

File: stdlib/public/Cxx/CxxRandomAccessCollection.swift [hunk 3]
```diff
@@ -0,0 +1,65 @@
...
+extension CxxRandomAccessCollection {
+  public var startIndex: Int {
+    return 0
+  }
+
+  public var endIndex: Int {
+    return count
+  }
+
+  public var count: Int {
```
Comment (hyp) [inline stdlib/public/Cxx/CxxRandomAccessCollection.swift:51]: make this inlinable?

File: stdlib/public/Cxx/CxxRandomAccessCollection.swift [hunk 4]
```diff
@@ -0,0 +1,65 @@
...
+
+  /// Do not implement this function manually in Swift.
+  func __beginUnsafe() -> RawIterator
+
+  /// Do not implement this function manually in Swift.
+  func __endUnsafe() -> RawIterator
+}
+
+extension CxxRandomAccessCollection {
+  public var startIndex: Int {
```
Comment (hyp) [inline stdlib/public/Cxx/CxxRandomAccessCollection.swift:43]: make this inlinable?

**Comment by egorzhdan** [comment]
@swift-ci please smoke test

File: stdlib/public/Cxx/CxxRandomAccessCollection.swift (hunk 1 above)
Comment (egorzhdan) [reply stdlib/public/Cxx/CxxRandomAccessCollection.swift]: Yeap, thanks

**Comment by egorzhdan** [comment]
@swift-ci please smoke test Linux

**Pull request merged**